src/context/values.js

```diff
@@ -0,0 +1,720 @@
+   * value. Several values for the same property can be set on a node - one
+   * per each setter. A repeated call for the same setter overwrites a
+   * previously set input. This is similar to how the same CSS property can be
+   * set in a stylesheet using different classes. The property computation
+   * decides how to treat several values. For instance:
+   * - A property can pick the first set input value (e.g. `font-size`).
+   * - A property can pick the min/max set input value (e.g. `priority`).
+   * - A property can reduce all values using `AND` (e.g. `renderable`).
+   *
+   * Once the input is set, the recalculation is rescheduled asynchronously.
+   * All dependent properties are also recalculated.
+   *
+   * @param {!ContextProp<T>} prop
+   * @param {*} setter
+   * @param {T} value
+   * @template T
+   */
+  set(prop, setter, value) {
+    devAssert(setter);
+    devAssert(value !== undefined);
+
+    const {key} = prop;
+
+    const inputsByKey = this.inputsByKey_ || (this.inputsByKey_ = new Map());
+    let inputs = inputsByKey.get(key);
+    if (!inputs) {
+      inputs = {
+        values: [],
+        setters: [],
+      };
+      inputsByKey.set(key, inputs);
+    }
+    const index = inputs.setters.indexOf(setter);
+    const changed = index == -1 || inputs.values[index] !== value;
+    if (index == -1) {
+      inputs.setters.push(setter);
+      inputs.values.push(value);
+    } else if (changed) {
+      inputs.values[index] = value;
+    }
+
+    if (changed) {
+      // An input has been added to a node for a first time. This might
+      // affect all values in this and child nodes. The simplest algorithm
+      // here is to deep scan all descendants and refresh them.
+      // Optimization opportunity: in simple but common manipulations the
+      // deepscan can be avoided.
+      this.ping(prop, false);
+      if (isRecursive(prop)) {
+        deepScan(this.contextNode_, scan, prop, true, false);
+      }
+    }
+  }
+
+  /**
+   * Unsets the input value for the specified property and setter.
+   * See `set()` for more info.
+   * @param {!ContextProp} prop
+   * @param {*} setter
+   */
+  remove(prop, setter) {
+    devAssert(setter);
+
+    const {key} = prop;
+    const inputsByKey = this.inputsByKey_;
+    const inputs = inputsByKey && inputsByKey.get(key);
+    if (inputs) {
+      const index = inputs.setters.indexOf(setter);
+      if (index != -1) {
+        inputs.setters.splice(index, 1);
+        inputs.values.splice(index, 1);
+        if (inputs.setters.length == 0) {
+          inputsByKey.delete(key);
+        }
+        deepScan(this.contextNode_, scan, prop);
+      }
+    }
+  }
+
+  /**
+   * Whether this node has inputs for the specified property.
+   *
+   * @param {!ContextProp} prop
+   * @return {boolean}
+   */
+  has(prop) {
+    const inputsByKey = this.inputsByKey_;
+    return inputsByKey && inputsByKey.has(prop.key);
```

Comment:
```suggestion
    return !!inputsByKey && inputsByKey.has(prop.key);
```

Reply:
Surprised the compiler didn't yell about this.